### ISSUE TITLE
stat.gcodes out of sync with parameter[5220] -> wrong DRO read out

### ISSUE THREAD

**Issue by Sigma1912** [LinuxCNC/linuxcnc#2745]
I'm reissuing this one with a clear description:

Using a recent build of version 2.9.1 

1. Start the 'axis' sim config
2. Home the machine
3. Load an .ngc like this one:
```
g10 l2 p2 x1
g55
g0 x10
m2
```
4. Run the above gcode
5. We get an error message about limit violations which is to be expected however stat.gcodes (used to display the 'Active G-Codes') and the value of parameter[5220] (which I presume is used for the DRO) are now out of sync:
 
![axis_out_of_sync](https://github.com/LinuxCNC/linuxcnc/assets/46067220/eeb010ac-4c11-4dc6-9f20-e916488e09e2)

My reaction would now be to simply dismiss the 'Active G-Codes' and argue that the DRO is what counts. However:

6. Run MDI command: 'G0 X0'
7.  If the DRO were correct we would expect the x axis to move from ' X:  -1.0000' to 'X: 0.0000' but nothing changes.
8.  Run MDI command: 'G0 X1' 
9.  Instead of moving to X : 1.0000 the DRO changes from' X:  -1.0000' to 'X: 0.0000'

**Comment by Sigma1912** [comment]
Actually I just noticed that switching offset index with emccanon commands can lead to inconsistencies not only with the index value but also with the offset values as apparently it is not always the values stored in the numbered parameters that are used but others that are stored elsewhere. I guess in the gui until that gets closed.

**Comment by Sigma1912** [comment]
I clearly didn't understand the way things work so I close this.

**Comment by Sigma1912** [comment]
Reopened with a clear bug description

**Comment by Sigma1912** [comment]
Workaround:

With this in the [RS274NGC]-section of the ini:

ON_ABORT_COMMAND = o <on_abort> call


and this in the on_abort.ngc:

```

o<on_abort> sub
g59
g54
o<on_abort> endsub
m2
```

Notes: 
- We need to switch to an offset-index other than g54 first for this to work (eg 'g59' as in the example above)
- Putting an 'm2' seems to also work, that however is something we are explicitly told not to do:

+
![m2 in oword](https://github.com/LinuxCNC/linuxcnc/assets/46067220/3b337e26-ada5-4cd9-abcd-202568606eb6)


https://linuxcnc.org/docs/devel/html/remap/remap.html#cha:remap

**Comment by rmu75** [comment]
> It seems that an error during gcode execution will reset the stat.gcodes to G54 but not parameter[5220].

How did you check the parameter?

I think the "current" WCS is somewhat poorly defined in the interpreter. For reasons of path optimization, the interpreter is running way ahead of where the machine is, so the GUI has to look at the state tags of the motion currently executing. If there is no motion happening then it's (probably) the EMC_STAT where info comes from. Maybe in this error case some event that DRO / backplot needs updating is lost (in the GUIs), but it may also be possible something else is at fault.

**Comment by Sigma1912** [comment]
> How did you check the parameter?


Basically what I found is:

sometimes 'self.stat.gcodes' disagrees with 'self.status.g5x_index'
and it seems 'self.status.g5x_index' reflects parameter[5220]
while 'self.stat.gcodes' gets updated in some other way.

When this happens the position values in the DRO do not reflect the internal state anymore.



Background:

The reason I came across this behavior is that I wanted to use emccanon. commands in a python remap to change the current WCS like this:
```

                self.params[5220] = 9
                emccanon.SET_G5X_OFFSET(
                    9,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                )
                yield INTERP_EXECUTE_FINISH
```

What I noticed is that  in AxisGUI the DRO will change the WCS to 'G59.3 ' (ie 9) however the Active Gcodes (ie stat.gcodes) is not updated to the new WCS (ie it does not change to 'G59.3'). 

In Gmoccapy the WCS index in the DRO follows the one in 'Active GCodes' so I  investigated this line 247  in 'combi_dro.py' (used in gmoccapy):

        self.gstat.connect('user-system-changed', self._user_system_changed)

which is indeed executed when the above emccanon. command runs and eventually executes this code that uses status.gcodes to display the current WCS:

```
    # get the actual coordinate system to display it on the DRO
    def _get_current_system(self):
            gcode = self.status.gcodes[1:]
            for code in gcode:
                if code >= 540 and code <= 590:
                    return "G%s" % int((code / 10))
                elif code > 590 and code <= 593:
                    return "G%s" % int((code / 10.0))
            return "Rel" 
```

So basically 'Gstat' correctly sends the message that there has been a change in the WCS system but because this change is not reflected in self.status.gcodes the DRO does not change.

Note that if we use 'status.g5x_index' in the above code:

           code = self.status.g5x_index

 the Gmoccapy DRO will then also disagree with the 'Active Gcodes' which leads me to presume that 'status.g5x' reflects the state of parameter[5220].




Only later did I realize that this state can also occur when not using python remaps and when this occurs  that the position in the DRO in both Axis and Gmoccapy and likely all the other Guis is also wrong.

Unfortunately I'm usually quickly lost as soon as C/C++ code is involved which seems to be where  the 'status' channel is coming from.

**Comment by rmu75** [comment]
IIRC  'self.stat.gcodes' comes from motion state tags, and that is only valid while some sort of motion is happening.

The whole code around that is kinda messy and has probably bugs. On the gui side some unification is needed, now there are too many parallel implementations of basic logic which are (probably) slightly different.

**Comment by Sigma1912** [comment]
Looking at this, it seems to me that parameter values may only get updated through gcodes (and m2?):
![parameters_set_by_gcode](https://github.com/LinuxCNC/linuxcnc/assets/
46067220/7e83d3fc-9b26-4f0f-8023-fdae1b7fb275)

Is this correct?
Which would maybe explain why a reset on abort would leave the parameter values out of sync with the system


On the other hand we have two 'SET_G5x_OFFSET' functions:
![set_g5x_offset](https://github.com/LinuxCNC/linuxcnc/assets/46067220/4313ba05-4d23-440c-ba15-f2cb861a9f5b)

Is one of these where the 'self.stat' tags come from?
If so what is the other?

**Comment by Sigma1912** [comment]
> If there is no motion happening then it's (probably) the EMC_STAT where info comes from.

Not sure if that is what you meant by 'EMC_STAT' but it's not coming from 'emcStatus->task.g92_offset.tran.x' (I can change that value in halui.cc and the DRO does not change it's behavior at all)

**Comment by Sigma1912** [comment]
Further testing has shown that the problem seems to come from 'm2' at the end of the program.
Using the legacy % format makes the bug disappear:

```
%
g10 l2 p2 x1
g55
;(abort,abort)
g0 x1000000000

%
```

Testing with '(abort, abort message)' instead of the limit violation leaves 'gcodes' with 'G55' and the DRO with 'G54':

![Image](https://github.com/user-attachments/assets/b5f68c77-65ed-404e-b79f-08033b6a5660)

**Comment by Sigma1912** [comment]
Seems the underlying issue is that the interpreter is running this part of the code in 'emc/rs274ngc/interp_convert.cc' and thus resets the parameters but due to the abort the corresponding 'PROGRAM_STOP()' it adds to the command queue is never actually executed.

![Image](https://github.com/user-attachments/assets/29e08f2a-1038-4372-9b6e-805aa3317ff5)

Removing the highlighted code above makes the bug disappear (but it will then also not do a reset on a 'm2' that is actually executed)
